src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Migrations/24.diff.sql

```diff
@@ -0,0 +1,146 @@
+/*************************************************************
+    This migration removes the existing primary key clustered index and adds a clustered index on Id column in the ResourceChangeData table.
+    The migration is "online" meaning the server is fully available during the upgrade, but it can be very time-consuming.
+    For reference, a resource change data table with 10 million records took around 25 minutes to complete 
+    on the Azure SQL database (SQL elastic pools - GeneralPurpose: Gen5, 2 vCores).
+**************************************************************/
+
+EXEC dbo.LogSchemaMigrationProgress 'Beginning migration to version 24.';
+GO
+
+EXEC dbo.LogSchemaMigrationProgress 'Adding or updating FetchResourceChanges_3 stored procedure.';
+GO
+
+--
+-- STORED PROCEDURE
+--     FetchResourceChanges_3
+--
+-- DESCRIPTION
+--     Returns the number of resource change records from startId. The start id is inclusive.
+--
+-- PARAMETERS
+--     @startId
+--         * The start id of resource change records to fetch.
+--     @lastProcessedUtcDateTime
+--         * The last checkpoint datetime in UTC time (Coordinated Universal Time).
+--     @pageSize
+--         * The page size for fetching resource change records.
+--
+-- RETURN VALUE
+--     Resource change data rows.
+--
+CREATE OR ALTER PROCEDURE dbo.FetchResourceChanges_3
+    @startId bigint,
+    @lastProcessedUtcDateTime datetime2(7),
+    @pageSize smallint
+AS
+BEGIN
+   
+    SET NOCOUNT ON;
+
+    /* Finds the prior partition to the current partition where the last processed watermark lies. It is a normal scenario when a prior watermark exists. */
+    DECLARE @precedingPartitionBoundary AS datetime2(7) = (SELECT TOP(1) CAST(prv.value as datetime2(7)) AS value FROM sys.partition_range_values AS prv WITH (NOLOCK)
```

Comment:
I think with the changes we no longer need to check the prior watermark partition.  Only the current partitions and the next.  Was there specific scenario the prior one was needed for in the latest testing?

Reply:
I believe that there is a scenario of a record would be inserted into the prior partition, with an identity value that is greater than the current watermark id.  For example:

Let's assume the last processed id is 49. Now, the current watermark will be 50.

Id | Timestamp | Partition number
49 2021-12-10 12:00:00.2000008             1
51 2021-12-10 11:59:59.9999988             1
50 2021-12-10 12:00:00.2222222             2